voxblox_ros/src/tsdf_server.cc

```diff
@@ -187,39 +212,59 @@ void TsdfServer::processPointCloudMessageAndInsert(
     const sensor_msgs::PointCloud2::Ptr& pointcloud_msg,
     const Transformation& T_G_C, const bool is_freespace_pointcloud) {
   // Convert the PCL pointcloud into our awesome format.
-  // TODO(helenol): improve...
+
   // Horrible hack fix to fix color parsing colors in PCL.
+  bool color_pointcloud = false;
   for (size_t d = 0; d < pointcloud_msg->fields.size(); ++d) {
     if (pointcloud_msg->fields[d].name == std::string("rgb")) {
       pointcloud_msg->fields[d].datatype = sensor_msgs::PointField::FLOAT32;
+      color_pointcloud = true;
     }
   }
 
-  pcl::PointCloud<pcl::PointXYZRGB> pointcloud_pcl;
-  // pointcloud_pcl is modified below:
-  pcl::fromROSMsg(*pointcloud_msg, pointcloud_pcl);
-
-  timing::Timer ptcloud_timer("ptcloud_preprocess");
-
   Pointcloud points_C;
   Colors colors;
-  points_C.reserve(pointcloud_pcl.size());
-  colors.reserve(pointcloud_pcl.size());
-  for (size_t i = 0; i < pointcloud_pcl.points.size(); ++i) {
-    if (!std::isfinite(pointcloud_pcl.points[i].x) ||
-        !std::isfinite(pointcloud_pcl.points[i].y) ||
-        !std::isfinite(pointcloud_pcl.points[i].z)) {
-      continue;
-    }
+  timing::Timer ptcloud_timer("ptcloud_preprocess");
 
-    points_C.push_back(Point(pointcloud_pcl.points[i].x,
-                             pointcloud_pcl.points[i].y,
-                             pointcloud_pcl.points[i].z));
-    colors.push_back(
-        Color(pointcloud_pcl.points[i].r, pointcloud_pcl.points[i].g,
-              pointcloud_pcl.points[i].b, pointcloud_pcl.points[i].a));
+  // Convert differently depending on RGB or I type.
+  if (color_pointcloud) {
+    pcl::PointCloud<pcl::PointXYZRGB> pointcloud_pcl;
+    // pointcloud_pcl is modified below:
+    pcl::fromROSMsg(*pointcloud_msg, pointcloud_pcl);
+    points_C.reserve(pointcloud_pcl.size());
+    colors.reserve(pointcloud_pcl.size());
+    for (size_t i = 0; i < pointcloud_pcl.points.size(); ++i) {
+      if (!std::isfinite(pointcloud_pcl.points[i].x) ||
```

Comment:
I feel the if (color_pointcloud) could go inside the loop and save a lot of code double up

Reply:
Couldn't do this because it needs to be converted to different PCL datatypes `<PointXYZRGB>`  vs `<PointXYZI>` and I couldn't figure out how to do this any nicer :(